caravel/assets/spec/javascripts/explore/components/actions_spec.js

```diff
@@ -1,10 +1,109 @@
-import { expect } from 'chai';
+import configureMockStore from 'redux-mock-store';
+import thunk from 'redux-thunk';
+import nock from 'nock';
+import { expect, assert } from 'chai';
+import sinon from 'sinon';
 import shortid from 'shortid';
 import * as actions from '../../../../javascripts/explorev2/actions/exploreActions';
 import { initialState } from '../../../../javascripts/explorev2/stores/store';
 import { exploreReducer } from '../../../../javascripts/explorev2/reducers/exploreReducer';
 
+const middlewares = [thunk];
+const mockStore = configureMockStore(middlewares);
+
+describe('ajax call for datasource metadata', () => {
+  afterEach(() => {
+    nock.cleanAll();
+  });
+
+  it('should return a function', () => {
+    expect(actions.setFormOpts(999, 'test')).to.be.function;
+  });
+
+  it('should dispatch clearAllOpts', () => {
+    const dispatch = sinon.spy();
+    actions.setFormOpts(null, null)(dispatch);
+    assert(dispatch.withArgs(actions.clearAllOpts()).calledOnce);
+  });
+
+  it('should dispatch new opts', () => {
+    const params = ['datasource_id=999', 'datasource_type=test'];
+    nock('/caravel/')
+      .get('fetch_datasource_metadata?' + params.join('&'))
+      .reply('success',
+      {
+        datasource_class: 'SqlaTable',
+        time_columns: [],
+        time_grains: [],
+        groupby_cols: [],
+        metrics: [],
+        filter_cols: [],
+      });
+
+    const store = mockStore(initialState);
+    store.dispatch = sinon.spy();
+    store.dispatch(actions.setFormOpts(999, 'test'));
+    expect(store.dispatch.callCount).to.equal(5);
```

Comment:
I'm having trouble figuring this one out. I looked at multiple docs for testing ajax action creators, sinon fakeServer, nock etc.  I think [http://redux.js.org/docs/recipes/WritingTests.html](url) seems to make more sense. I kind of modified some of it, but this test keeps failing. Saying the actual callCount of dispatch is 1 instead of 5. I figure the dispatch callcount may not take callbacks into account. Still trying to figure this out.